docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,423 @@
+* If an operator anticipates a scenario of permanent quorum loss, they can trigger an `on-demand snapshot compaction` to create a compacted full-snapshot. This can potentially reduce the recovery time from a permanent quorum loss.
+
+##### Task Config
+
+We do not need any config for this task. When creating an instance of `EtcdOperatorTask` for this scenario, `.spec.config` will be set to nil (unset).
+
+##### Pre-Conditions
+
+* There should not be a `on-demand snapshot compaction` task already running for the same etcd cluster.
+
+> Note: `on-demand snapshot compaction` runs as a separate job in a separate pod, which interacts with the backup bucket and not the etcd cluster itself, hence it doesn't depend on the health of etcd cluster members.
+
+#### Trigger on-demand full/delta snapshot
+
+`Etcd` custom resource provides an ability to set [FullSnapshotSchedule](https://github.com/gardener/etcd-druid/blob/master/api/v1alpha1/types_etcd.go#L158) which currently defaults to run once in 24 hrs. [DeltaSnapshotPeriod](https://github.com/gardener/etcd-druid/blob/master/api/v1alpha1/types_etcd.go#L171) is also made configurable which defines the duration after which a delta snapshot will be taken.
+If a human operator does not wish to wait for the scheduled full/delta snapshot, they can trigger an on-demand (out-of-schedule) full/delta snapshot on the etcd cluster, which will be taken by the `leading-backup-restore`.
+
+##### Possible scenarios
+
+* An on-demand full snapshot can be triggered if scheduled snapshot fails due to any reason.
+* [Gardener Shoot Hibernation](https://github.com/gardener/gardener/blob/master/docs/usage/shoot_hibernate.md): Every etcd cluster incurs an inherent cost of preserving the volumes even when a gardener shoot control plane is scaled down, i.e the shoot is in a hibernated state. However, it is possible to save on hyperscaler costs by invoking this task to take a full snapshot before scaling down the etcd cluster, and deleting the etcd data volumes afterwards.
+* [Control Plane Migration](https://github.com/gardener/gardener/blob/master/docs/proposals/07-shoot-control-plane-migration.md): In [gardener](https://github.com/gardener/gardener), a cluster control plane can be moved from one seed cluster to another. This process currently requires the etcd data to be replicated on the target cluster, so a full snapshot of the etcd cluster in the source seed before the migration would allow for faster restoration of the etcd cluster in the target seed.
+
+##### Task Config
+
+```go
+// SnapshotType can be full or delta snapshot.
+type SnapshotType string
+
+const (
+  FullSnapshot SnapshotType = "full-snapshot"
+  DeltaSnapshot SnapshotType = "delta-snapshot"
+)
+
+type OnDemandSnapshotTaskConfig struct {
+  metav1.TypeMeta   `json:",inline"`
+  // Type of on-demand snapshot.
+  Type SnapshotType `json:"type"`
+}
+```
+
+```yaml
+spec:
+  config:
+    apiversion: druid.gardener.cloud/v1alpha1
```

Comment:
we do not need to have `apiversion` and `kind` as part of the config. Instead of using `RawExtension` just change the type to `string` and use YAML literal operator to specify a yaml as the value for config.
Example:
```yaml
spec:
  config: |
    type: <type of on-demand snapshot>
```
Can you make this change elsewhere as well?